### PR TITLE
[docs] Improved inpaint docs

### PR DESCRIPTION
**Pull request by stevhliu**
Part of #4758 to update the inpainting guide to be more extensive and in-depth

### CODE REVIEW

File: docs/source/en/using-diffusers/inpaint.md [hunk 1]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
-# Text-guided image-inpainting
+# Inpainting
 
 [[open-in-colab]]
 
-The [`StableDiffusionInpaintPipeline`] allows you to edit specific parts of an image by providing a mask and a text prompt. It uses a version of Stable Diffusion, like [`runwayml/stable-diffusion-inpainting`](https://huggingface.co/runwayml/stable-diffusion-inpainting) specifically trained for inpainting tasks.
+Inpainting replaces or edits specific areas of an image. This makes it a useful tool for image restoration like removing defects and artifacts, or even replacing an image area with something entirely new. Inpainting relies on a mask to determine which regions of an image to fill in; the area to inpaint is represented by white pixels and the area to keep is represented by black pixels. The white pixels are filled in by the prompt.
 
-Get started by loading an instance of the [`StableDiffusionInpaintPipeline`]:
+With 🤗 Diffusers, here is how you can quickly start inpainting:
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:19]:
```suggestion
With 🤗 Diffusers, here is how you can do inpainting:
```

File: docs/source/en/using-diffusers/inpaint.md [hunk 2]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+  </div>
+</div>
 
-```python
-def download_image(url):
-    response = requests.get(url)
-    return PIL.Image.open(BytesIO(response.content)).convert("RGB")
+## Popular models
 
+[Stable Diffusion v1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5), [Stable Diffusion XL (SDXL)](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), and [Kandinsky 2.2](https://huggingface.co/kandinsky-community/kandinsky-2-2-decoder-inpaint) are among the most popular models for inpainting. SDXL typically produces higher resolution images than Stable Diffusion v1.5, and Kandinsky 2.2 is also capable of generating high-quality images thanks to an image prior model.
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:69]:
```suggestion
[Stable Diffusion Inpainting](https://huggingface.co/runwayml/stable-diffusion-inpainting), [Stable Diffusion XL (SDXL) Inpainting](diffusers/stable-diffusion-xl-1.0-inpainting-0.1), and [Kandinsky 2.2](https://huggingface.co/kandinsky-community/kandinsky-2-2-decoder-inpaint) are among the most popular models for inpainting. SDXL typically produces higher resolution images than Stable Diffusion v1.5, and Kandinsky 2.2 is also capable of generating high-quality images thanks to an image prior model.
```

File: docs/source/en/using-diffusers/inpaint.md [hunk 3]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
 
-img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
-mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"
+### Stable Diffusion v1.5
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:71]:
```suggestion
### Stable Diffusion Inpainting
```

File: docs/source/en/using-diffusers/inpaint.md [hunk 4]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
 
-init_image = download_image(img_url).resize((512, 512))
-mask_image = download_image(mask_url).resize((512, 512))
+Stable Diffusion v1.5 is a latent diffusion model finetuned on 512x512 images. It is a good starting point for inpainting because it is relatively fast and generates good quality images. To use this model for inpainting, you'll need to pass a prompt, base and mask image to the pipeline:
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:73]:
```suggestion
Stable Diffusion Inpainting is a latent diffusion model finetuned on 512x512 images on inpainting. It is a good starting point for inpainting because it is relatively fast and generates good quality images. To use this model for inpainting, you'll need to pass a prompt, base and mask image to the pipeline:
```

File: docs/source/en/using-diffusers/inpaint.md [hunk 5]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
-image = pipeline(prompt=prompt, image=init_image, mask_image=mask_image).images[0]
+SDXL is a larger and more powerful version of Stable Diffusion v1.5. This model can follow a two-stage model process (though each model can also be used alone); the base model generates an image, and a refiner model takes that image and further enhances its details and quality. Take a look at the [SDXL](sdxl) guide for a more comprehensive guide on how to use SDXL and configure it's parameters.
+
+```py
+import torch
+from diffusers import AutoPipelineForInpainting
+from diffusers.utils import load_image
+
+pipeline = AutoPipelineForInpainting.from_pretrained(
+    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16"
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:105]: We should use this checkpoint here: diffusers/stable-diffusion-xl-1.0-inpainting-0.1

File: docs/source/en/using-diffusers/inpaint.md [hunk 6]
```diff
@@ -121,4 +320,247 @@ unmasked_unchanged_image = PIL.Image.fromarray(unmasked_unchanged_image_arr.roun
...
+    <figcaption class="mt-2 text-center text-sm text-gray-500">text-to-image</figcaption>
+  </div>
+  <div class="flex-1">
+    <img class="rounded-xl" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/inpaint-text-chain-out.png"/>
+    <figcaption class="mt-2 text-center text-sm text-gray-500">inpaint</figcaption>
+  </div>
+</div>
+
+
+### Image-to-image-to-inpaint
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/inpaint.md:379]: It's mostly the other way that is interesting here:
- a) First do inpainting
- b) Then do img2img or upscaling, e.g. to improve quality

**Review by patrickvonplaten — COMMENTED** [review]
Nice start, left some comments

File: docs/source/en/using-diffusers/img2img.md [hunk 7]
```diff
@@ -590,17 +590,17 @@ image
...
+Running diffusion models is computationally expensive and intensive, but with a few optimization tricks, it is entirely possible to run them on consumer and free-tier GPUs. For example, you can use a more memory-efficient form of attention such as PyTorch 2.0's [scaled-dot product attention](../optimization/torch2.0#scaled-dot-product-attention) or [xFormers](../optimization/xformers) (you can use one or the other, but there's no need to use both). You can also offload the model to the GPU while the other pipeline components wait on the CPU.
 
 ```diff
 + pipeline.enable_model_cpu_offload()
 + pipeline.enable_xformers_memory_efficient_attention()
 ```
 
-With [`torch.compile`](optimization/torch2.0#torch.compile), you can boost your inference speed even more by wrapping your UNet with it:
+With [`torch.compile`](../optimization/torch2.0#torch.compile), you can boost your inference speed even more by wrapping your UNet with it:
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/img2img.md:601]: I feel like we're using this block of text across many guides now. How can we focus more on the objectives of the guides themselves and redirect users for the optimization related details, instead? 

(Doesn't need to be addressed in this PR and I am open to brainstorming)

File: docs/source/en/using-diffusers/inpaint.md (hunk 2 above)
Comment (sayakpaul) [reply docs/source/en/using-diffusers/inpaint.md:69]: I don't think "... thanks to an image prior model" is adding any additional info here. I think we can discard that bit. No strong opinions.

File: docs/source/en/using-diffusers/inpaint.md [hunk 8]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+init_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/inpaint.png").convert("RGB")
+mask_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/inpaint_mask.png").convert("RGB")
+
+generator = torch.Generator("cuda").manual_seed(92)
+prompt = "concept art digital painting of an elven castle, inspired by lord of the rings, highly detailed, 8k"
+image = pipeline(prompt=prompt, image=init_image, mask_image=mask_image, generator=generator).images[0]
 ```
 
-Now you can create a prompt to replace the mask with something else:
+### Stable Diffusion XL (SDXL)
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/inpaint.md:95]: Should this be SDXL Inpainting following what we're doing in the rest of the guide? 

* SDXL stands for Stable Diffusion XL and that's already mentioned above.
* We're using "Stable Diffusion Inpainting". So, I guess it makes sense to use "SDXL Inpainting" as well?

File: docs/source/en/using-diffusers/inpaint.md [hunk 9]
```diff
@@ -10,87 +10,286 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+
+generator = torch.Generator("cuda").manual_seed(92)
+prompt = "concept art digital painting of an elven castle, inspired by lord of the rings, highly detailed, 8k"
+image = pipeline(prompt=prompt, image=init_image, mask_image=mask_image, generator=generator).images[0]
 ```
 
-`image`          | `mask_image` | `prompt` | output |
-:-------------------------:|:-------------------------:|:-------------------------:|-------------------------:|
-<img src="https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png" alt="drawing" width="250"/> | <img src="https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png" alt="drawing" width="250"/> | ***Face of a yellow cat, high resolution, sitting on a park bench*** | <img src="https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/in_paint/yellow_cat_sitting_on_a_park_bench.png" alt="drawing" width="250"/> |
+### Kandinsky 2.2
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/inpaint.md:119]: Kandinsky 2.2 Inpainting?

File: docs/source/en/using-diffusers/inpaint.md [hunk 10]
```diff
@@ -121,4 +320,247 @@ unmasked_unchanged_image = PIL.Image.fromarray(unmasked_unchanged_image_arr.roun
 unmasked_unchanged_image.save("force_unmasked_unchanged.png")
 ```
 
-Forcing the unmasked portion of the image to remain the same might result in some weird transitions between the unmasked and masked areas, since the model will typically change the masked and unmasked areas to make the transition more natural.
+## Chained inpainting pipelines
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/inpaint.md:323]: Love this!

File: docs/source/en/using-diffusers/inpaint.md [hunk 11]
```diff
@@ -121,4 +320,247 @@ unmasked_unchanged_image = PIL.Image.fromarray(unmasked_unchanged_image_arr.roun
...
+
+[`AutoPipelineForInpainting`] can be chained with other 🤗 Diffusers pipelines to edit their outputs.
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/inpaint.md:325]: Should we also add a line or two about how chained pipelines are beneficial?

File: docs/source/en/using-diffusers/inpaint.md [hunk 12]
```diff
@@ -121,4 +320,247 @@ unmasked_unchanged_image = PIL.Image.fromarray(unmasked_unchanged_image_arr.roun
...
++ pipeline.enable_model_cpu_offload()
+```
+
+To speed-up your inference code even more, use [`torch_compile`](../optimization/torch2.0#torch.compile). You should wrap `torch.compile` around the most intensive component in the pipeline which is typically the UNet:
+
+```py
+pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)
+```
+
+Learn more in the [Reduce memory usage](../optimization/memory) and [Torch 2.0](../optimization/torch2.0) guides.
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/inpaint.md:566]: Yeah after this section, I definitely think we should think a bit more about how we can organize and distribute these components that are heavily reused across different docs. 

(Doesn't have to be addressed in this PR).

**Review by sayakpaul — COMMENTED** [review]
Thank you for the changes!

File: docs/source/en/using-diffusers/inpaint.md (hunk 5 above)
Comment (stevhliu) [reply docs/source/en/using-diffusers/inpaint.md:105]: No wonder my results looked off! 😭 I'll update

File: docs/source/en/using-diffusers/inpaint.md (hunk 10 above)
Comment (yiyixuxu) [inline docs/source/en/using-diffusers/inpaint.md:323]: question for @patrickvonplaten

I remember I did this PR https://github.com/huggingface/diffusers/pull/4819 to supports chaining inpainting with other pipelines but it's not obvious how though -  we need to create `masked_image_latents` from `image` and `mask` with an additional step no?

**Review by yiyixuxu — APPROVED** [review]
awesome, thank you!

File: docs/source/en/using-diffusers/inpaint.md (hunk 10 above)
Comment (patrickvonplaten) [reply docs/source/en/using-diffusers/inpaint.md:323]: If you change the mask then yes, but if you keep the same mask then I don't think it's necessary. 

One might want to always keep the same mask, but inpaint multiple times

File: src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py [hunk 13]
```diff
@@ -426,7 +426,10 @@ def encode_prompt(
             self._lora_scale = lora_scale
 
             # dynamically adjust the LoRA scale
-            adjust_lora_scale_text_encoder(self.text_encoder, lora_scale)
+            if not self.use_peft_backend:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/unidiffuser/pipeline_unidiffuser.py:429]: Is this related?

File: src/diffusers/schedulers/scheduling_unipc_multistep.py [hunk 14]
```diff
@@ -282,13 +282,13 @@ def _threshold_sample(self, sample: torch.FloatTensor) -> torch.FloatTensor:
         https://arxiv.org/abs/2205.11487
         """
         dtype = sample.dtype
-        batch_size, channels, *remaining_dims = sample.shape
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_unipc_multistep.py:285]: Is this related? (think it's fixed on main btw :-))

**Review by patrickvonplaten — APPROVED** [review]
Nice that works for me - it would be great though to make sure that we only change .md files here :-)
